--- ../../../base/lua/plugins/lsp.lua	2024-04-08 13:26:52.632240285 +0800
+++ ./lsp[mason].lua	2024-04-18 09:46:59.920517072 +0800
@@ -3,6 +3,8 @@
     "neovim/nvim-lspconfig",
     event = "VeryLazy",
     dependencies = {
+      "williamboman/mason.nvim",
+      "williamboman/mason-lspconfig.nvim",
       "hrsh7th/nvim-cmp",
       "hrsh7th/cmp-nvim-lsp",
       'hrsh7th/cmp-nvim-lsp-signature-help',
@@ -14,6 +16,23 @@
       'L3MON4D3/LuaSnip',
     },
     config = function ()
+      require("mason").setup()
+      require("mason-lspconfig").setup {
+        ensure_installed = {
+          -- LSPs kept installed without extra operations
+        },
+        -- Auto install LSPs
+        automatic_installation = false,
+        -- Handlers
+        handlers = {
+          function (lsp_id)
+            require("lspconfig")[lsp_id].setup {
+              capabilities = require('cmp_nvim_lsp').default_capabilities(),
+            }
+          end,
+        },
+      }
+
       local cmp = require('cmp')
       cmp.setup {
         snippet = {
@@ -51,28 +70,6 @@
 
       local capabilities = require('cmp_nvim_lsp').default_capabilities()
 
-      local function default_config(t)
-        for _, name in ipairs(t) do
-          lspconfig[name].setup {
-            capabilities = capabilities,
-          }
-        end
-      end
-
-      default_config({
-        -- LSP name here for default configuration
-        --
-        -- See https://github.com/neovim/nvim-lspconfig/blob/master/doc/server_configurations.md
-        -- for all options.
-        --
-        -- E.g.
-        --
-        -- "bashls",
-        -- "clangd",
-        -- "rust_analyzer",
-        -- "zls",
-      })
-
       -- Lua
       lspconfig.lua_ls.setup {
         on_init = function(client)
