--- ./base/lua/plugins/lsp.lua	2024-04-18 23:41:24.667824115 +0800
+++ ./optional/lua/plugins/lsp[mason].lua	2024-04-18 23:41:12.761127116 +0800
@@ -3,6 +3,8 @@
     "neovim/nvim-lspconfig",
     event = "VeryLazy",
     dependencies = {
+      "williamboman/mason.nvim",
+      "williamboman/mason-lspconfig.nvim",
       "hrsh7th/nvim-cmp",
       "hrsh7th/cmp-nvim-lsp",
       'hrsh7th/cmp-nvim-lsp-signature-help',
@@ -14,6 +16,23 @@
       'L3MON4D3/LuaSnip',
     },
     config = function ()
+      require("mason").setup()
+      require("mason-lspconfig").setup {
+        ensure_installed = {
+          -- LSPs kept installed without extra operations
+        },
+        -- Auto install LSPs
+        automatic_installation = false,
+        -- Handlers
+        handlers = {
+          function (lsp_id)
+            require("lspconfig")[lsp_id].setup {
+              capabilities = require('cmp_nvim_lsp').default_capabilities(),
+            }
+          end,
+        },
+      }
+
       local cmp = require('cmp')
       cmp.setup {
         snippet = {
@@ -61,6 +80,7 @@
 
       default_config({
         -- LSP names here for default configuration
+        -- (Only for LSPs not managed by Mason)
         --
         -- See https://github.com/neovim/nvim-lspconfig/blob/master/doc/server_configurations.md
         -- for all options.
